File: Lib/Vector3.lua
local Utils = require("Lib.Utils")

local setmetatable = setmetatable
local type = type
local rawget = rawget
local m_sqrt = math.sqrt

local GetUnitX = GetUnitX
local GetUnitY = GetUnitY

---@class Vector3
local cls = {}

cls._loc = Location(0, 0)

local function getTerrainZ(x, y)
    MoveLocation(cls._loc, x, y)
    return GetLocationZ(cls._loc)
end

---@return Vector3
function cls.new(x, y, z)
    x = x or 0
    y = y or 0
    return setmetatable({
        x = x,
        y = y,
        z = z or getTerrainZ(x, y),
    }, cls)
end

local new = cls.new

---@param unit unit
function cls.FromUnit(unit)
    local x = GetUnitX(unit)
    local y = GetUnitY(unit)
    return new(x, y, getTerrainZ(x, y) + GetUnitFlyHeight(unit))
end

--function cls.InsideUnitCircle()
--    local angle = math.random() * math.pi * 2
--    return new(math.cos(angle), math.sin(angle))
--end

---@param unit unit
function cls:MoveToUnit(unit)
    self.x = GetUnitX(unit)
    self.y = GetUnitY(unit)
    self.z = getTerrainZ(self.x, self.y) + GetUnitFlyHeight(unit)
    return self
end

---@param unit unit
---@param mode integer modes. 1: force flying. 2: force to ground. other|Nil: flying units fly/ ground units grounded
function cls:UnitMoveTo(unit, mode)
    local tz = getTerrainZ(self.x, self.y)
    local defaultFlyHeight = GetUnitDefaultFlyHeight(unit)
    local minZ = tz + defaultFlyHeight
    SetUnitPosition(unit, self.x, self.y)
    if mode == 1 then
        Utils.SetUnitFlyable(unit)
        SetUnitFlyHeight(unit, math.max(minZ, self.z) - minZ, 0)
    elseif mode == 2 then
        SetUnitFlyHeight(unit, defaultFlyHeight, 0)
    else
        if IsUnitType(unit, UNIT_TYPE_FLYING) then
            SetUnitFlyHeight(unit, math.max(minZ, self.z) - minZ, 0)
        else
            SetUnitFlyHeight(unit, defaultFlyHeight, 0)
        end
    end
    return self
end

---@param other Vector3
function cls:SetTo(other)
    self.x = other.x
    self.y = other.y
    self.z = other.z
    return
end

---@param other Vector3
function cls:Add(other)
    self.x = self.x + other.x
    self.y = self.y + other.y
    self.z = self.z + other.z
    return self
end

---@param other Vector3
function cls:Sub(other)
    self.x = self.x - other.x
    self.y = self.y - other.y
    self.z = self.z - other.z
    return self
end

---@param d real
function cls:Div(d)
    self.x = self.x / d
    self.y = self.y / d
    self.z = self.z / d
    return self
end

---@param d real
function cls:Mul(d)
    self.x = self.x * d
    self.y = self.y * d
    self.z = self.z * d
    return self
end

function cls:SetNormalize()
    local magnitude = self:GetMagnitude()

    if magnitude > 1e-05 then
        self:Div(magnitude)
    else
        self.x = 0
        self.y = 0
        self.z = 0
    end

    return self
end

function cls:SetMagnitude(len)
    self:SetNormalize():Mul(len)
    return self
end

function cls:Clone()
    return new(self.x, self.y, self.z)
end

function cls:GetTerrainZ()
    return getTerrainZ(self.x, self.y)
end

function cls:GetMagnitude()
    return m_sqrt(self:SqrMagnitude())
end

function cls:SqrMagnitude()
    return self.x * self.x + self.y * self.y + self.z * self.z
end

function cls.Dot(lhs, rhs)
    return lhs.x * rhs.x + lhs.y * rhs.y + lhs.z * rhs.z
end

function cls.Scale(a, b)
    local x = a.x * b.x
    local y = a.y * b.y
    local z = a.z * b.z
    return new(x, y, z)
end

function cls.Cross(lhs, rhs)
    local x = lhs.y * rhs.z - lhs.z * rhs.y
    local y = lhs.z * rhs.x - lhs.x * rhs.z
    local z = lhs.x * rhs.y - lhs.y * rhs.x
    return new(x, y, z)
end

function cls.Project(v, onNormal)
    local num = onNormal:SqrMagnitude()

    if num < 0.0001 then
        return new(0, 0, 0)
    end

    local num2 = cls.Dot(v, onNormal)
    local v3 = onNormal:Clone()
    v3:Mul(num2 / num)
    return v3
end

function cls.ProjectOnPlane(v, planeNormal)
    local v3 = cls.Project(v, planeNormal)
    v3:Mul(-1)
    v3:Add(v)
    return v3
end

---@return string
function cls:tostring()
    return string.format("(%f,%f,%f)", self.x, self.y, self.z)
end

function cls.__index(_, k)
    return rawget(cls, k)
end

function cls.__add(a, b)
    return new(a.x + b.x, a.y + b.y, a.z + b.z)
end

---@return Vector3
function cls.__sub(a, b)
    return new(a.x - b.x, a.y - b.y, a.z - b.z)
end

function cls.__div(v, d)
    return new(v.x / d, v.y / d, v.y / d)
end

function cls.__mul(a, d)
    if type(d) == "number" then
        return new(a.x * d, a.y * d, a.z * d)
    else
        return a:Clone():MulQuaternion(d)
    end
end

function cls.__unm(v)
    return new(-v.x, -v.y, -v.z)
end

function cls.__eq(a, b)
    return ((a.x - b.x) ^ 2 + (a.y - b.y) ^ 2 + (a.z - b.z) ^ 2) < 9.999999e-11
end

function cls.up()
    return new(0, 0, 1)
end

function cls.down()
    return new(0, 0, -1)
end

function cls.right()
    return new(1, 0, 0)
end

function cls.left()
    return new(-1, 0, 0)
end

function cls.forward()
    return new(0, 1, 0)
end

function cls.back()
    return new(0, -1, 0)
end

function cls.zero()
    return new(0, 0, 0)
end

function cls.one()
    return new(1, 1, 1)
end

setmetatable(cls, cls)

return cls

File: CSProject/Utils/Vector3.cs
using LuaWrapper;
using SFLib.Interop;

public enum UnitVec3Mode
{
    ForceFlying,
    ForceGround,
    /// <summary>
    /// Flying units fly, ground units grounded.
    /// </summary>
    Auto,
}

public struct Vector3
{
    private static location _loc = Location(0, 0);

#region Static Properties

    public static Vector3 zero => new(0, 0, 0);
    public static Vector3 up => new(0, 0, 1);
    public static Vector3 down => new(0, 0, -1);
    public static Vector3 right => new(1, 0, 0);
    public static Vector3 left => new(-1, 0, 0);
    public static Vector3 forward => new(0, 1, 0);
    public static Vector3 back => new(0, -1, 0);
    public static Vector3 one => new(1, 1, 1);

#endregion

#region Operators

    public static Vector3 operator +(Vector3 a, Vector3 b)
    {
        return new Vector3(a.x + b.x, a.y + b.y, a.z + b.z);
    }

    public static Vector3 operator -(Vector3 a)
    {
        return new Vector3(-a.x, -a.y, -a.z);
    }

    public static Vector3 operator -(Vector3 a, Vector3 b)
    {
        return new Vector3(a.x - b.x, a.y - b.y, a.z - b.z);
    }

    public static Vector3 operator *(Vector3 v, float f)
    {
        return new Vector3(v.x * f, v.y * f, v.z * f);
    }

    public static Vector3 operator *(float f, Vector3 v)
    {
        return new Vector3(v.x * f, v.y * f, v.z * f);
    }

    public static Vector3 operator /(Vector3 v, float f)
    {
        return new Vector3(v.x / f, v.y / f, v.z / f);
    }

#endregion

    /// <summary>
    /// Warcraft III world space here is right-handed: +x points right, +y points away, +z points up.
    /// That means Cross((1,0,0), (0,1,0)) == (0,0,1).
    /// </summary>
    public static Vector3 Cross(Vector3 a, Vector3 b)
    {
        return new Vector3(a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x);
    }

    public static float Distance(Vector3 a, Vector3 b)
    {
        return (a - b).magnitude;
    }

    public static float Dot(Vector3 a, Vector3 b)
    {
        return a.x * b.x + a.y * b.y + a.z * b.z;
    }

    public static Vector3 Lerp(Vector3 a, Vector3 b, float t)
    {
        t = math.clamp01(t);
        return a + (b - a) * t;
    }

    public static Vector3 MoveTowards(Vector3 current, Vector3 target, float maxDistanceDelta)
    {
        var toVector = target - current;
        var dist = toVector.magnitude;
        if (dist <= maxDistanceDelta || dist == 0f)
        {
            return target;
        }
        return current + toVector / (dist / maxDistanceDelta);
    }

    public static Vector3 Project(Vector3 v, Vector3 onNormal)
    {
        var sqrMag = Dot(onNormal, onNormal);
        if (sqrMag < 0.0001f)
        {
            return zero;
        }
        var dot = Dot(v, onNormal);
        return onNormal * (dot / sqrMag);
    }

    public static Vector3 ProjectOnPlane(Vector3 v, Vector3 planeNormal)
    {
        return v - Project(v, planeNormal);
    }

    public static Vector3 Reflect(Vector3 inDirection, Vector3 inNormal)
    {
        return inDirection - 2 * Dot(inDirection, inNormal) * inNormal;
    }

    public static Vector3 RotateTowards(Vector3 current, Vector3 target, float maxRadiansDelta, float maxMagnitudeDelta)
    {
        var currentMag = current.magnitude;
        var targetMag = target.magnitude;

        if (currentMag == 0f || targetMag == 0f)
        {
            return MoveTowards(current, target, maxMagnitudeDelta);
        }

        var currentNorm = current / currentMag;
        var targetNorm = target / targetMag;

        var dot = math.clamp(Dot(currentNorm, targetNorm), -1f, 1f);
        var angle = math.acos(dot);

        if (angle == 0f)
        {
            return MoveTowards(current, target, maxMagnitudeDelta);
        }

        var t = math.min(1f, maxRadiansDelta / angle);
        var newDir = Slerp(currentNorm, targetNorm, t);
        var newMag = math.moveTowards(currentMag, targetMag, maxMagnitudeDelta);
        return newDir * newMag;
    }

    public static Vector3 Scale(Vector3 a, Vector3 b)
    {
        return new Vector3(a.x * b.x, a.y * b.y, a.z * b.z);
    }

    public static Vector3 Slerp(Vector3 a, Vector3 b, float t)
    {
        var magA = a.magnitude;
        var magB = b.magnitude;

        if (magA == 0f || magB == 0f)
        {
            return MoveTowards(a, b, math.huge);
        }

        var normA = a / magA;
        var normB = b / magB;

        var dot = math.clamp(Dot(normA, normB), -1f, 1f);
        var angle = math.acos(dot);
        var sinAngle = math.sin(angle);

        if (sinAngle < 0.0001f)
        {
            return MoveTowards(a, b, math.huge);
        }

        var tAngle = angle * t;
        var sinTA = math.sin(tAngle);
        var sinTOneMinusA = math.sin(angle - tAngle);

        var newDir = (normA * sinTOneMinusA + normB * sinTA) / sinAngle;
        var newMag = math.lerp(magA, magB, t);
        return newDir * newMag;
    }

    private static float _getTerrainZ(float x, float y)
    {
        MoveLocation(_loc, x, y);
        return GetLocationZ(_loc);
    }

    public static Vector3 FromUnit(unit u)
    {
        var x = GetUnitX(u);
        var y = GetUnitY(u);
        return new Vector3(x, y, _getTerrainZ(x, y) + GetUnitFlyHeight(u));
    }

    public float x;
    public float y;
    public float z;

    public float sqrMagnitude => x * x + y * y + z * z;
    public float magnitude => math.sqrt(sqrMagnitude);

    public Vector3 normalized
    {
        get
        {
            var mag = magnitude;
            if (mag < 0.0001f)
            {
                return zero;
            }
            return this / mag;
        }
    }

    public Vector3(float x, float y, float z)
    {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3 ClampMagnitude(float mag)
    {
        return normalized * mag;
    }

    public override string ToString()
    {
        return $"({x}, {y}, {z})";
    }

    public void UnitMoveTo(unit u, UnitVec3Mode mode = UnitVec3Mode.Auto)
    {
        var tz = _getTerrainZ(x, y);
        var defaultFlyHeight = GetUnitDefaultFlyHeight(u);
        var minZ = tz + defaultFlyHeight;
        SetUnitPosition(u, x, y);
        switch (mode)
        {
            case UnitVec3Mode.ForceFlying:
                LuaUtils.SetUnitFlyable(u);
                SetUnitFlyHeight(u, math.max(minZ, z) - minZ, 0f);
                break;
            case UnitVec3Mode.ForceGround:
                SetUnitFlyHeight(u, defaultFlyHeight, 0f);
                break;
            case UnitVec3Mode.Auto:
                if (IsUnitType(u, UNIT_TYPE_FLYING))
                {
                    SetUnitFlyHeight(u, math.max(minZ, z) - minZ, 0f);
                }
                else
                {
                    SetUnitFlyHeight(u, defaultFlyHeight, 0f);
                }
                break;
        }
    }

    public float GetTerrainZ()
    {
        return _getTerrainZ(x, y);
    }
}
